### ISSUE TITLE
Bug: camelcase v8.0.0 regression on properties: "never"

### ISSUE THREAD

**Issue by GuillemotF** [eslint/eslint#15572]
### Environment

Node version: 16.7.0
npm version: 7.20.3
Local ESLint version: 8.8.0
Global ESLint version: none
Operating System: Ubuntu

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "rules": {
        "camelcase": [
            "error",
            { "properties": "never", "ignoreDestructuring": true }
        ]
    }
}
```
</details>

```js
var { some_property } = obj;
doSomething({ some_property });
```

### What did you expect to happen?

No lint error as described in the last statement of https://eslint.org/docs/rules/camelcase#ignoredestructuring-true

### What actually happened?

Identifier 'some_property' is not in camel case. (line 2)

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

This is working as intended until v7.32.0
Not working since 8.0.0 release.

**Comment by snitin315** [comment]
Looks like a bug, since it is given as a correct lint-free example in the docs.

```js
/*eslint camelcase: ["error", {"properties": "never", ignoreDestructuring: true}]*/

var { some_property } = obj;
doSomething({ some_property });
```

### [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgY2FtZWxjYXNlOiBbXCJlcnJvclwiLCB7XCJwcm9wZXJ0aWVzXCI6IFwibmV2ZXJcIiwgaWdub3JlRGVzdHJ1Y3R1cmluZzogdHJ1ZX1dKi9cblxudmFyIHsgc29tZV9wcm9wZXJ0eSB9ID0gb2JqO1xuZG9Tb21ldGhpbmcoeyBzb21lX3Byb3BlcnR5IH0pOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsiZ2xvYmFsUmV0dXJuIjp0cnVlfX0sInJ1bGVzIjp7InF1b3RlcyI6Mn0sImVudiI6e319fQ==)

Though, the following code is lint-free - 

```js
/*eslint camelcase: ["error", {"properties": "never", ignoreDestructuring: true}]*/

var { some_property } = obj;
doSomething(some_property);
```

### [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgY2FtZWxjYXNlOiBbXCJlcnJvclwiLCB7XCJwcm9wZXJ0aWVzXCI6IFwibmV2ZXJcIiwgaWdub3JlRGVzdHJ1Y3R1cmluZzogdHJ1ZX1dKi9cblxudmFyIHsgc29tZV9wcm9wZXJ0eSB9ID0gb2JqO1xuZG9Tb21ldGhpbmcoc29tZV9wcm9wZXJ0eSk7XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7Imdsb2JhbFJldHVybiI6dHJ1ZX19LCJydWxlcyI6eyJxdW90ZXMiOjJ9LCJlbnYiOnt9fX0=)

#


Also, I did not find any test case present for this.

**Comment by mdjermanovic** [comment]
> ```js
> var { some_property } = obj;
> doSomething({ some_property });
> ```
>
> Identifier 'some_property' is not in camel case. (line 2)

Given that this rule reports all references to non-camelcased variables, this behavior is actually logical.

In `var { some_property } = obj`, shorthand `some_property` is a property name used to read property value (which is always allowed by this rule) and a variable name in destructuring pattern (which is allowed by `"ignoreDestructuring": true`).

In `doSomething({ some_property })`, shorthand `some_property` is a property name used to create property (which is allowed by `"properties": "never"`) and a variable name. It is reported as a variable name (a reference to the variable). because this isn't a destructuring pattern, so `"ignoreDestructuring": true` doesn't apply here.

> This is working as intended until v7.32.0
> Not working since 8.0.0 release.

It seems that ESLint v7.32.0 had a bug - `"properties": "never"` should allow property names, but this option was mistakenly allowing all child nodes of Property nodes:

```js
// ESLint v7.32.0

/* eslint camelcase: ["error", { ignoreDestructuring: true, properties: "never" }] */

const foo_bar = 5; // reported.

foo_bar; // reported as a reference to variable.

const obj = {
    a: foo_bar, // not reported. It's a bug because this is a reference to variable, not a property name.
    [foo_bar]: 6, // not reported. It's a bug because this is a reference to variable, not a property name.
    foo_bar,  // not reported. It's a bug because this is both a reference to variable and a property name.
}
```

This bug has been fixed in ESLint v8:

```js
// ESLint v8.8.0

/* eslint camelcase: ["error", { ignoreDestructuring: true, properties: "never" }] */

const foo_bar = 5; // reported.

foo_bar; // reported as a reference to variable.

const obj = {
    a: foo_bar, // reported as a reference to variable.
    [foo_bar]: 6, // reported as a reference to variable.
    foo_bar,  // reported as a reference to variable.
}
```

That said, the example from the original post was one of the two primary use cases for `ignoreDestructuring` option, so I agree that we should change the rule to allow this example for backward compatibility, although the current behavior makes more sense.
